src/coreComponents/dataRepository/DataContext.hpp

```diff
@@ -0,0 +1,270 @@
+/// - its location in the data hierarchy, implementation in GroupContext and WrapperContext.
+/// Typically, the target object contains an unique_ptr< DataContext > instance of this class.
+class DataContext
+{
+public:
+
+  /**
+   * @brief Construct a new DataContext object.
+   * @param objectName the target object name
+   * @param isDataFileContext true if this Context is a DataFileContext
+   */
+  DataContext( string const & objectName, bool isDataFileContext );
+
+  /**
+   * @brief Destroy the DataContext object
+   */
+  virtual ~DataContext() {}
+
+  /**
+   * @return A string that mention all the known informations to retrieve from where the target
+   * object comes from.
+   */
+  virtual string toString() const = 0;
+
+  /**
+   * @brief Get the target object name
+   */
+  string getObjectName() const
+  { return m_objectName; }
+
+  /**
+   * @brief Flag on availability of file information. Used to provide more user-friendly information.
+   * @return true if the context is from a file.
+   */
+  bool isDataFileContext() const
+  { return m_isDataFileContext; }
+
+  /**
+   * @brief Insert contextual information in the provided stream.
+   */
+  friend std::ostream & operator<<( std::ostream & os, const DataContext & dt );
+
+protected:
+
+  /// see getObjectName()
+  string const m_objectName;
+
+  /// see isDataFileContext()
+  bool const m_isDataFileContext;
+
+};
+
+/// Helps to know where a Group is in the hierarchy.
+/// See DataContext class for more info.
+class GroupContext : public DataContext
+{
+public:
+
+  /**
+   * @brief Construct a new GroupContext object
+   * @param group The reference to the Group related to this GroupContext.
+   */
+  GroupContext( Group & group );
+
+  /**
+   * @brief Destroy the GroupContext object
+   */
+  virtual ~GroupContext() {}
+
+  /**
+   * @return the reference to the Group related to this GroupContext.
+   */
+  Group & getGroup() const;
+
+  /**
+   * @return the group path with the file & line of the first parent for which this information exists.
+   */
+  virtual string toString() const;
+
+protected:
+
+  /**
+   * @brief Construct a new GroupContext object
+   * @param group The reference to the Group related to this GroupContext.
+   * @param objectName Target object name.
+   */
+  GroupContext( Group & group, string const & objectName );
+
+  /// The reference to the Group related to this GroupContext.
+  Group & m_group;
+
+};
+
+/// Dedicated implementation of GroupContext for Wrapper.
+/// See DataContext class for more info.
+class WrapperContext final : public GroupContext
+{
+public:
+
+  /**
+   * @brief Construct a new WrapperContext object
+   * @param wrapper the target Wrapper object
+   */
+  WrapperContext( WrapperBase & wrapper );
+
+  /**
+   * @brief Destroy the WrapperContext object
+   */
+  virtual ~WrapperContext() {}
+
+  /**
+   * @return the parent group DataContext followed by the wrapper name.
+   */
+  virtual string toString() const;
```

Comment:
Would something like https://en.cppreference.com/w/cpp/language/cast_operator be relevant?

Reply:
I hesitated to use it but it seemed like a bad practice. Implicit casts can sometimes lead to unintentional implicit conversions, and explicit casts would not really be shorter here.